### PR TITLE
Speed up dependency resolution

### PR DESCRIPTION
**Pull request by oehme**
### Context
I was looking to speed up dependency resolution in a large project and found some low-hanging fruit. 

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/excludes/simple/DefaultExcludeAllOf.java [hunk 1]
```diff
@@ -44,20 +44,33 @@ protected String getDisplayName() {
...
 
     @Override
     public boolean mayExcludeArtifacts() {
         if (mayExcludeArtifacts != null) {
             return mayExcludeArtifacts;
         }
-        mayExcludeArtifacts = components().allMatch(ExcludeSpec::mayExcludeArtifacts);
+        mayExcludeArtifacts = true;
+        for (ExcludeSpec component : getComponents()) {
+            mayExcludeArtifacts &= component.mayExcludeArtifacts();
```
Comment (bamboo) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/excludes/simple/DefaultExcludeAllOf.java:72]: 💡 consider breaking out of the loop early once `component.mayExcludeArtifacts()` returns false

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/excludes/simple/DefaultExcludeAnyOf.java [hunk 2]
```diff
@@ -44,20 +44,33 @@ protected String getDisplayName() {
...
 
     @Override
     public boolean mayExcludeArtifacts() {
         if (mayExcludeArtifacts != null) {
             return mayExcludeArtifacts;
         }
-        mayExcludeArtifacts = components().anyMatch(ExcludeSpec::mayExcludeArtifacts);
+        mayExcludeArtifacts = false;
+        for (ExcludeSpec component : getComponents()) {
+            mayExcludeArtifacts |= component.mayExcludeArtifacts();
```
Comment (bamboo) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/excludes/simple/DefaultExcludeAnyOf.java:72]: 💡 consider breaking out of the loop early once `component.mayExcludeArtifacts()` returns true

**Comment by big-guy** [comment]
@bot-gradle test this

**Comment by oehme** [comment]
@bamboo https://builds.gradle.org/viewLog.html?buildId=48540562&buildTypeId=Gradle_Master_Check_ConfigCache_20_Trigger seems to be a bug in the config cache. It can end up putting half-initialized objects into Sets when that Object is encountered again during its own initialization. This is common in dependency graphs. It worked previously because ResolvedDependencyResult used `Object.hashCode` which can be called at any point.

Do you think you could fix this bug or should I work around it?

**Comment by bamboo** [comment]
@bot-gradle test ACC

**Comment by bamboo** [comment]
Hey @oehme, thanks for uncovering this limitation in the configuration cache.

https://github.com/gradle/gradle/pull/20493 is the best we could come up with in time for 7.5. It doesn't fully allow your optimizations yet as problems would still be reported during serialization* but perhaps we could find a way to selectively silence the problems for the types in this PR (assuming all circular references are safe)?

Could you rebase this PR on top of master and see what you get? 

* in the general case Gradle can't really know a safe order to initialize circular references as that would require tracking `hashCode` implementation dependencies

**Comment by ljacomet** [comment]
Closing this one in favor of #20555 which drops the problematic change.

**Comment by oehme** [comment]
Sorry for the late response. I don't think the last commit was worth an elaborate workaround. All of these were pretty small improvements. Thanks for merging the rest @ljacomet

**Pull request closed without merging**